### ISSUE TITLE
Extend SelfEquality and ComparisonOutOfRange to handle </>/<=/>=

### ISSUE THREAD

**Issue by cpovirk** [google/error-prone#274]
I thought I had results for running this check over Google code (at least for inputs of type `int`), but I can't find them. I'm re-running now.

**Comment by kluever** [comment]
See also https://github.com/JodaOrg/joda-time/pull/681

**Comment by cpovirk** [comment]
The SelfEquality end of this was handled when that check was generalized into IdentityBinaryExpression.

The ComparisonOutOfRange end sounds like the kind of thing for me to play with in my last 10 minutes before my long weekend... :)

**Comment by cpovirk** [comment]
The 10-minute investigation made things look easy, but I have been racking my brain as I try to figure out how to finish and polish it :) I should be able to get there, but I'm going to try to at least somewhat back-burner this.

So far, I've found a lot of `0 <= charValue && charValue < somethingElse` checks, which is redundant but harmless code. (If I'm feeling up to it, maybe I'll produce a nicer suggested edit for that case. Really, the current suggested edits (usually to replace the expression with `true` or `false`) aren't typically very useful except for testing purposes.) I also found `someLong < 2E48`, which was meant to be `someLong < (1<<48)`, so that was a nice find, albeit in code that's been dead for a few years and probably wouldn't have hurt anything, anyway. I think there were some other hits, but I just meant to give a quick update to say that I'll try to get back to this.

The main other thing I want to look into is looking for checks that aren't against a pure literal, like against `Character.MAX_VALUE` or `(1 << 32) - 1`. The former might be my canonical example of this problem: `for (char c = 0; c <= Character.MAX_VALUE; c++)`. That may bring some false positives (perhaps especially in generated code?), but I'm hoping that that's manageable.
